_posts/2018-09-20-rails-security.markdown

```diff
@@ -0,0 +1,96 @@
+---
+layout: post
+title:  "Step up the security of your Rails app | Part 1"
+date: 2018-09-20 12:10:00
+categories: ["rails", "security"]
+author: "lubc"
+---
+
+The internet is a wonderful place, but there will always be people that don't have good intentions when they visit our websites. That's why you need to be aware of the vulnerabilities that your application can have and how to avoid them. In this article I'll cover two common security problems in [Rails](https://rubyonrails.org) applications (I'll probably make a second part since this is a very extensive topic).
+
+<!--more-->
+
+It's also worth to mention that Rails itself has improved a lot [over the years](https://www.youtube.com/watch?v=Btrmc1wO3pc) to make everything more secure and easy for the people who use it. That's one of the reasons why it's so important to [upgrade your Rails project](https://www.ombulabs.com/blog/tags/upgrades).
+
+Before we start you should know that Rails has an [official security guide](https://guides.rubyonrails.org/security.html) with a large variety of common vulnerabilities. I recommend to take a look at it if you want to dive deeper into this topic. Also at the end of this article I'll leave a few links to some useful tools that will help you to level up your security in Rails.
+
+Okay, let's jump into it:
+
+## Command Injection
+#### What is it?
+It allows the attacker to run any command on your server.
+[Ruby](https://www.ruby-lang.org/en/) has a [method](https://apidock.com/ruby/Kernel/eval) called `eval` which can take a string and execute what's inside of it. This can likely be the reason of this security issue if you use it wrong.
+
+#### Example
+
+```ruby
+# User input
+params[:shop][:items_ids] # Maybe you expect this to be an array inside a string.
+                          # But it can contain something very dangerous like:
+                          # "Kernel.exec('Whatever OS command you want')"
+
+# Vulnerable code
+evil_string = params[:shop][:items_ids]
+eval(evil_string)
+
+# BOOM!
+```
+
+#### How to avoid it?
+`eval` is a very powerful method that should only be used in a few specific cases. Most of the time it is possible to accomplish the same goal with a safer solution.
+A rule of thumb is that you should never use it with user input. At least without properly sanitizing it.
```

Comment:
@lubc Maybe this sounds better? 

"If you see a call to `eval` you must be very sure that you are properly sanitizing it."

Reply:
Also, maybe you can add an example of good sanitization?